// dojo-shim/dist/umd/native/number.js
(function (factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        var v = factory(require, exports); if (v !== undefined) module.exports = v;
    }
    else if (typeof define === 'function' && define.amd) {
        define(["require", "exports", '../support/global'], factory);
    }
})(function (require, exports) {
    "use strict";
    var global_1 = require('../support/global');
    /**
     * The smallest interval between two representable numbers.
     */
    exports.EPSILON = global_1.default.Number.EPSILON;
    /**
     * The maximum safe integer in JavaScript
     */
    exports.MAX_SAFE_INTEGER = global_1.default.Number.MAX_SAFE_INTEGER;
    /**
     * The minimum safe integer in JavaScript
     */
    exports.MIN_SAFE_INTEGER = global_1.default.Number.MIN_SAFE_INTEGER;
    /**
     * Determines whether the passed value is NaN without coersion.
     *
     * @param value The value to test
     * @return true if the value is NaN, false if it is not
     */
    exports.isNaN = global_1.default.Number.isNaN;
    /**
     * Determines whether the passed value is a finite number without coersion.
     *
     * @param value The value to test
     * @return true if the value is finite, false if it is not
     */
    exports.isFinite = global_1.default.Number.isFinite;
    /**
     * Determines whether the passed value is an integer.
     *
     * @param value The value to test
     * @return true if the value is an integer, false if it is not
     */
    exports.isInteger = global_1.default.Number.isInteger;
    /**
     * Determines whether the passed value is an integer that is 'safe,' meaning:
     *   1. it can be expressed as an IEEE-754 double precision number
     *   2. it has a one-to-one mapping to a mathematical integer, meaning its
     *      IEEE-754 representation cannot be the result of rounding any other
     *      integer to fit the IEEE-754 representation
     *
     * @param value The value to test
     * @return true if the value is an integer, false if it is not
     */
    exports.isSafeInteger = global_1.default.Number.isSafeInteger;
});
